File: 477. Total Hamming Distance/main.cpp
#include <bitset>
#include <vector>

using std::bitset;
using std::vector;

class Solution {
public:
    int totalHammingDistance(vector<int> &nums) {
        int total = 0;
        int size = nums.size();

        for (int i = 0; i < 32; i++) {
            int bitCount = 0;
            for (int j = 0; j < size; j++) {
                bitCount += (nums[j] >> i) & 1;
            }
            total += bitCount * (size - bitCount);
        }

        return total;
    }
};

class Solution {
public:
    int totalHammingDistance(vector<int> &nums) {
        int total = 0;
        int size = nums.size();

        for (int i = 0; i < size - 1; i++) {
            for (int j = i + 1; j < size; j++) {
                total += bitset<32>(nums[i] ^ nums[j]).count();
            }
        }

        return total;
    }
};

File: 477. Total Hamming Distance/main.rs
struct Solution;

impl Solution {
    pub fn total_hamming_distance(nums: Vec<i32>) -> i32 {
        let mut total = 0;
        let size = nums.len();

        for i in 0..32 {
            let mut bit_count = 0;
            for j in 0..size {
                bit_count += (nums[j] >> i) & 1;
            }
            total += bit_count * (size as i32 - bit_count);
        }

        total
    }
}

impl Solution {
    pub fn total_hamming_distance(nums: Vec<i32>) -> i32 {
        let mut total = 0;
        let size = nums.len();

        for i in 0..size - 1 {
            for j in i + 1..size {
                total += (nums[i] ^ nums[j]).count_ones();
            }
        }

        total as i32
    }
}

impl Solution {
    pub fn total_hamming_distance(nums: Vec<i32>) -> i32 {
        let mut bit_count = [0; 32];
        let size = nums.len();

        nums.iter().for_each(|n| {
            (0..32).for_each(|i| {
                bit_count[i] += (n >> i) & 1;
            })
        });

        bit_count
            .iter()
            .map(|count| count * (size as i32 - count))
            .sum()
    }
}
